# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFT2FontBase.cpp
# Commit: a884b6f401c3
# Full Hash: a884b6f401c38b471e0fe4a82739d72180032697
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2025-05-09 04:15:00
# Regressor Bug: 1961387
# File Overlap Count: 1
# Description:
#   Bug 1961387 - Account for scaling of color-bitmap fonts when setting up the font-unit scale factor. r=gfx-reviewers,lsalzman
#   
#   This fixes the problem of color-bitmap Noto Color Emoji getting incorrect
#   glyph advances in vertical mode. Without the scaling correction here, we
#   end up with a constant glyph advance regardless of the font size, just
# ==============================================================================

diff -r eafaeb768277 -r a884b6f401c3 gfx/thebes/gfxFT2FontBase.cpp
--- a/gfx/thebes/gfxFT2FontBase.cpp	Thu May 08 17:42:11 2025 +0000
+++ b/gfx/thebes/gfxFT2FontBase.cpp	Thu May 08 17:42:11 2025 +0000
@@ -351,9 +351,10 @@
     // we can get units_per_EM from the 'head' table instead; otherwise,
     // we don't have a unitsPerEm value so we can't compute/use yScale or
     // mFUnitsConvFactor (x scale).
-    const TT_Header* head =
-        static_cast<TT_Header*>(FT_Get_Sfnt_Table(face, ft_sfnt_head));
-    if (head) {
+    if (const TT_Header* head =
+            static_cast<TT_Header*>(FT_Get_Sfnt_Table(face, ft_sfnt_head))) {
+      gfxFloat emUnit = head->Units_Per_EM;
+      mFUnitsConvFactor = ftMetrics.x_ppem / emUnit;
       // Bug 1267909 - Even if the font is not explicitly scalable,
       // if the face has color bitmaps, it should be treated as scalable
       // and scaled to the desired size. Metrics based on y_ppem need
@@ -366,9 +367,8 @@
         mMetrics.maxDescent *= adjustScale;
         mMetrics.maxAdvance *= adjustScale;
         lineHeight *= adjustScale;
+        mFUnitsConvFactor *= adjustScale;
       }
-      gfxFloat emUnit = head->Units_Per_EM;
-      mFUnitsConvFactor = ftMetrics.x_ppem / emUnit;
       yScale = emHeight / emUnit;
     }
   }
